sdks/go/pkg/beam/core/runtime/harness/harness.go

```diff
@@ -338,14 +347,36 @@ func (c *control) handleInstruction(ctx context.Context, req *fnpb.InstructionRe
 		c.cache.CompleteBundle(tokens...)
 
 		mons, pylds := monitoring(plan, store)
+		requiresFinalization := false
 		// Move the plan back to the candidate state
 		c.mu.Lock()
 		// Mark the instruction as failed.
 		if err != nil {
 			c.failed[instID] = err
 		} else {
-			// Non failure plans can be re-used.
-			c.plans[bdID] = append(c.plans[bdID], plan)
+			// Non failure plans should either be moved to the finalized state
+			// or to plans so they can be re-used.
+			expiration := plan.GetExpirationTime(ctx, string(instID))
+			if time.Now().Before(expiration) {
+				// TODO(damccorm) - we can be a little smarter about data structures here by
+				// by storing plans awaiting finalization in a heap. That way when we expire plans
+				// here its O(1) instead of O(n) (though adding/finalizing will still be O(logn))
```

Comment:
+1 Especially once Generic container types begin to drop with Go 1.18. And I agree that maintaining our own Sorted By Time Slices would be a bit irritating in the short term, and require it's own set of testing. Not hard (because the sort package does exist, just finicky to test all the edges).

I will let you know that in Streaming jobs, there are typically O(100s) if not O(1000s) of bundles coming through in very quick succession, so we'll need to keep an eye out in case finalization becomes a bottleneck, once we're able to load test it in earnest.